chemicaltoolbox/rdock/sort_filter.xml

```diff
@@ -1,38 +1,65 @@
     <command><![CDATA[
 
-sdsort -n $descending -s -f'$sort_field' -id'$name_field' '$input' |
-  sdfilter -f'\$_COUNT <= $top' -s'$name_field' |
-  sdsort -n $descending -f'$sort_field' > '$output'
+cat '$input'
+#if $filters
+| sdfilter
+#for $i, $f in enumerate($filters)
+-f'$f.filter'
+#end for
+#end if
+#if $name_field
+| sdsort -n $descending -s -f'$sort_field' -id'$name_field'
+| sdfilter -f'\$_COUNT <= $top' -s'$name_field'
+#end if
+#if $global_sort and $sort_field
+| sdsort -n $descending -f'$sort_field'
+#end if
+> '$output'
 
     ]]></command>
 
     <inputs>
         <param type="data" name="input" format="sdf" label="Molecules" help="Molecules in SDF format"/>
-        <param name="top" type="integer" value="1" label="Number of records to keep in output" help="Number of best scoring records to keep"/>
-        <param name="sort_field" type="text" label="Field to sort on" optional="false" help="Name of the field to sort records by">
+        <repeat name="filters" title="Filters">
+            <param name="filter" type="text" label="Filter expression" optional="true" help="Perl expression for filter">
+                <sanitizer>
+                    <valid initial="string.printable">
+                        <remove value="&apos;"/>
+                    </valid>
+                    <mapping initial="none"/>
+                </sanitizer>
+            </param>
+        </repeat>
+
+        <param name="sort_field" type="text" label="Field to sort on" optional="true" help="Name of the field to sort records by">
             <sanitizer>
                 <valid initial="string.printable">
                     <remove value="&apos;"/>
                 </valid>
                 <mapping initial="none"/>
             </sanitizer>
         </param>
+        <param name="descending" type="boolean" label="Sort descending" truevalue="-r" falsevalue="" checked="true"
+               help="Sort ascending or descending"/>
+        <param name="global_sort" type="boolean" label="Global sort" checked="true"
+               help="Sort all records in file after filtering (true) or just sort within the blocks identified by $name_field (false)"/>
+
         <param name="name_field" type="text" label="Grouping field name" optional="false" help="Name of the field to group records by (must be sequential)">
             <sanitizer>
                 <valid initial="string.printable">
                     <remove value="&apos;"/>
                 </valid>
                 <mapping initial="none"/>
             </sanitizer>
-        </param>    
-        <param name="descending" type="boolean" label="Sort descending" truevalue="-r" falsevalue="" checked="true"
-               help="Generate the name field (first line) for cases where this is empty"/>
+        </param>
+        <param name="top" type="integer" value="1" label="Number of records to keep in output" optional="false" help="Number of best scoring records to keep"/>
```

Comment:
The help text is confusing I think. Is this just returning 1 record by default?

Reply:
Using default values yes. Within each group of records (grouped by field $name_field) you end up with the $top best records (according to $sort_field and $descending).

Reply:
But field_name is also optional, would it make sense to also keep "top" then optional?

Reply:
True, but top is ignored is field_name is not defined.

Reply:
On the commandline, but the user does not know that ... maybe at least adding this to the help?
I guess the proper way would be to model it with a conditional.

Reply:
I made all params optional.